### DISCUSSION TITLE
Add new device class for binary sensors: exception

### DISCUSSION THREAD

**Discussion by elahd** [home-assistant/architecture#696] [Entity Models]
# Context

While creating the [NYC 311 Public Services Calendar](https://github.com/elahd/ha-nyc311) integration, I felt that I was making compromises on selecting device classes for the integration's binary sensors. This integration reports whether there are scheduling exceptions to New York City's public schools, sanitation collection (trash/recycling/composting), and parking rules. It uses generic (`None` class) binary sensors to report when things are normal (`off`) or suspended/closed (`on`).  (In NYC, alternate side parking and parking meter rules are suspended on holidays.)

~~I'm proposing a few binary sensor changes in multiple posts here that I believe are generic enough to be included in ha's core.~~ _(Just this proposal for now.)_

# Proposal

Add a new `exception` device class to binary sensors.

This can be used for any entity that can report an exception state that is not necessarily a "problem". That is, for NYC public schools, holiday closures are a scheduling exception.

## Existing Device Classes

There are a few existing device classes that seem like they would apply here, but would ultimately be strange to use:

1. `problem`: A problem is an exception with negative connotation, whereas `exception` itself is neutral.
2. `opening`: For something like the school sensor, I only want the sensor to be `on` for non-routine closures (i.e.: holidays and snow days). An `opening` sensor would be inadequate for exceptional closures as it would be `off` for weekends, showing school as `open` when it is technically `closed`

I also considered using the `running` sensor, but that just doesn't _feel_ right.

# Suggested Implementation

Add a new `exception` device class for binary sensors.

| Sensor State | State Text  | State Icon                                                                                                    | State Icon Name    | Note                                         |
| ------------ | ----------- | ------------------------------------------------------------------------------------------------------------- | ------------------ | -------------------------------------------- |
| `True`       | `Exception` | ![image](https://user-images.githubusercontent.com/466460/147678867-ce7eee6a-73ed-4b58-881b-75bf29051334.png) | `mdi:alert`        | Same as `power`/`safety`/`smoke` `on` state. |
| `False`      | `Normal`    | ![image](https://user-images.githubusercontent.com/466460/147678518-db75672f-09e4-41d5-a1a3-bfea19496b0a.png) | `mdi:check-circle` | Same as `moving` device class' `on` state.   |

# Consequences

See below:

- https://github.com/home-assistant/core/pull/57632
- https://github.com/home-assistant/frontend/pull/10268
- https://github.com/home-assistant/developers.home-assistant/pull/1084
- https://github.com/home-assistant/home-assistant.io/pull/19755

**Comment by thecode** [comment]
Looking at your suggested implementation, why not make it device class alert? I think that alert will fit more situations and still can fit as an exception

**Reply by elahd** [reply]
I like it -- that would work for me and does seem to fit more situations.

**Reply by KNXBroker** [reply]
Thumbs up for alert device class. Since a long time I have a similar issue with my wind, fire and rain alarm sensors. None of the existing device classes match.

**Reply by farmio** [reply]
What about existing device class `problem`? I think it would also fit.
@KNXBroker for rain I'd use `moisture` for fire maybe `heat` or `smoke`.

**Reply by thecode** [reply]
@farmio I am not in the details for the wind/fire sensors, but the suggested device class alert is not necessarily a problem, maybe it is something you need to look or to be aware of, but does not have to be negative or positive in it's meaning.

**Reply by elahd** [reply]
Ideally, we would be able to customize binary sensors from within a custom component more than we can now. Today, we can only easily customize icons by overloading the `icon` function. The remaining properties that are essential to binary sensors are the text fields in https://github.com/home-assistant/core/blob/dev/homeassistant/components/binary_sensor/translations/en.json along with the constants in the BinarySensorDeviceClass in https://github.com/home-assistant/core/blob/dev/homeassistant/components/binary_sensor/__init__.py.

It's extremely difficult to override these fields.

Binary sensors don't convey any "meaning" beyond simply reporting state. That is, the rest of the platform doesn't treat a `heat` binary sensor any differently than it does a `door`. Given that device classes only govern display features and not behavior, I think **an alternate solution to this problem would be just opening up the binary_sensor platform to easy one-off customization.**

I'm not sure of the protocol here. Should this be proposed in a separate discussion? Will a maintainer eventually chime in with an official decision based on this conversation?

**Comment by kongo09** [comment]
I just implemented a similar integration for a Hamburg school using a binary sensor. Stumbling over the exact same question with device class, I can very much relate to the proposal. `alert` would work as well. Even broader and maybe less alarming could be `notify` which would be anything that justifies non-urgent attention.

**Reply by okridgway** [reply]
Why not allow for a custom device binary sensor?

**Reply by elahd** [reply]
@okridgway It would be nice to be able to make a custom binary sensor device class. It technically can be done, but it's a lot of work. The necessary parts need to be easier to override.
